checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java

```diff
@@ -96,28 +107,310 @@ public LockTransfer createFlowTransferFunction(CFAbstractAnalysis<CFValue, LockS
+
+    @Override
+    protected void annotateImplicit(Tree tree, AnnotatedTypeMirror type,
+            boolean iUseFlow) {
+        super.annotateImplicit(tree, type, iUseFlow);
+
+        if (tree.getKind() == Kind.METHOD_INVOCATION) {
+        	// If a method's formal return type is annotated with @GuardSatisfied(index),
+        	// look for the first instance of @GuardSatisfied(index) in the method definition's receiver type or
+        	// formal parameters, retrieve the corresponding type of the actual parameter / receiver at the call site
+        	// (e.g. @GuardedBy("someLock") and replace the return type at the call site with this type.
+
+            MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
+            Pair<AnnotatedExecutableType, List<AnnotatedTypeMirror>> mfuPair = methodFromUse(methodInvocationTree);
+            AnnotatedExecutableType invokedMethod = mfuPair.first;
+
+            if (invokedMethod.getElement().getKind() != ElementKind.CONSTRUCTOR) {
+                AnnotatedTypeMirror methodDefinitionReturn = invokedMethod.getReturnType().getErased();
+
+                if (methodDefinitionReturn != null && methodDefinitionReturn.hasAnnotation(GuardSatisfied.class)) {
+                    int returnGuardSatisfiedIndex = AnnotationUtils.
+                            getElementValue(methodDefinitionReturn.getAnnotation(GuardSatisfied.class), "value", Integer.class, true);
+
+                    // @GuardSatisfied with no index defaults to index -1. Ignore instances of @GuardSatisfied with no index.
+                    // If a method is defined with a return type of @GuardSatisfied with no index, an error is reported by LockVisitor.visitMethod.
+                    if (returnGuardSatisfiedIndex != -1) {
+
+	                    // Find the receiver or first parameter whose @GS index matches that of the return type.
+	                    // Ensuring that the type annotations on distinct @GS parameters with the same index match at the call site is handled in LockVisitor.visitMethodInvocation
+
+	                    ExecutableElement invokedMethodElement = invokedMethod.getElement();
+	                    if (!ElementUtils.isStatic(invokedMethodElement) && !TreeUtils.isSuperCall(methodInvocationTree)) {
+	                    	AnnotatedTypeMirror methodDefinitionReceiver = invokedMethod.getReceiverType().getErased();
+	                        if (methodDefinitionReceiver != null && methodDefinitionReceiver.hasAnnotation(GuardSatisfied.class)) {
+	                        	int receiverGuardSatisfiedIndex = AnnotationUtils.
+	                                    getElementValue(methodDefinitionReceiver.getAnnotation(GuardSatisfied.class), "value", Integer.class, true);
+
+	                            if (receiverGuardSatisfiedIndex == returnGuardSatisfiedIndex) {
+	                                type.replaceAnnotation(getReceiverType(methodInvocationTree).getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE));
```

Comment:
This code is 7 conditions deep. Can you maybe break down this code into smaller chunks and factor them into separate methods?